contrib/win32/openssh/conhost_conpty.c

```diff
@@ -0,0 +1,150 @@
+* notice, this list of conditions and the following disclaimer.
+* 2. Redistributions in binary form must reproduce the above copyright
+* notice, this list of conditions and the following disclaimer in the
+* documentation and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+* OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+* IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+* NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+* THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#include "includes.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "log.h"
+
+#include "conhost_conpty.h"
+
+const unsigned int PTY_SIGNAL_RESIZE_WINDOW = 8u;
+
+// Return Value: 0 for success, -1 for failure
+int
+CreateConPty(const char *cmdline,
+	const unsigned short width,
+	const unsigned short height,
+	HANDLE const hInput,
+	HANDLE const hOutput,	
+	HANDLE* const conhost_pty_sighandle,
+	PROCESS_INFORMATION* const piPty)
+{
+	HANDLE signal_pipe_conhost_side;
+	SECURITY_ATTRIBUTES sa;
+	char system32_path[PATH_MAX] = { 0, };
+
+	memset(&sa, 0, sizeof(SECURITY_ATTRIBUTES));
+	sa.nLength = sizeof(sa);	
+	sa.lpSecurityDescriptor = NULL;	
+	sa.bInheritHandle = TRUE;
+	
+	CreatePipe(&signal_pipe_conhost_side, conhost_pty_sighandle, &sa, 0);
+	SetHandleInformation(*conhost_pty_sighandle, HANDLE_FLAG_INHERIT, 0);
+
+	char conhostCmdline[8191] = { 0, }; // msdn	
+	char *cmd_fmt = "%s\\conhost.exe --headless --width %d --height %d --signal 0x%x -- %s";
+	
+	if (!GetSystemDirectoryA(system32_path, PATH_MAX))
+		fatal("unable to retrieve system32 path");
+	
+	snprintf(conhostCmdline,
+		_countof(conhostCmdline),
+		cmd_fmt,
+		system32_path,
+		width,
+		height,
+		signal_pipe_conhost_side,
+		cmdline);
+	
+	STARTUPINFO si;
+	memset(&si, 0, sizeof(STARTUPINFO));
+	si.cb = sizeof(STARTUPINFOW);
+	si.hStdInput = hInput;
+	si.hStdOutput = hOutput;	
+	si.hStdError = hOutput;
+	si.dwFlags = STARTF_USESTDHANDLES | STARTF_USESIZE | STARTF_USECOUNTCHARS;
+
+	debug3("conhostcmdline:%s", conhostCmdline);
+	
+	if (0 == CreateProcess(NULL, conhostCmdline, NULL, NULL, TRUE, 0, NULL, NULL, &si, piPty)) {
+		debug("Failed to create the conhost pty, error:%d", GetLastError());
+		
+		return -1;
+	}
+
+	return 0;
+}
+
+// Return 1 if the resize succeeded, else 0.
+int
+SignalResizeWindow(const HANDLE conhost_pty_sighandle, const unsigned short width, const unsigned short height)
```

Comment:
This is not hooked up to change_window_size request from client. Intentional?

Reply:
It's hooked up with window change req from ssh client. This method is called in session_window_change_req().